flow/record/base.py

```diff
@@ -723,66 +789,73 @@ def open_path(path, mode, clobber=True):
-                fp = dctx.stream_reader(fp)
+            fp = open_stream(fp, mode)
     return fp
 
 
-def RecordAdapter(url, out, selector=None, clobber=True, **kwargs):
-    url = str(url or "")
-
+def RecordAdapter(url, out: bool, selector=None, clobber=True, **kwargs) -> AbstractWriter | AbstractReader:
     # Guess adapter based on extension
     ext_to_adapter = {
         ".avro": "avro",
         ".json": "jsonfile",
         ".jsonl": "jsonfile",
         ".csv": "csvfile",
     }
-    _, ext = os.path.splitext(url)
+    cls_stream = None
+    if not out and hasattr(url, "read") and hasattr(url, "seek"):
+        # This record adapter has received a file-like object.
+        # We just need to find the right adapter
+        # by peeking into the first few bytes.
 
-    adapter_scheme = ext_to_adapter.get(ext, "stream")
-    if "://" not in url:
-        url = f"{adapter_scheme}://{url}"
+        # First, we open the stream. If the stream is compressed, open_stream will wrap it for us into a decompressor.
+        cls_stream = Peekable(open_stream(url, "rb"))
 
-    p = urlparse(url, scheme=adapter_scheme)
-    adapter, _, sub_adapter = p.scheme.partition("+")
+        # We will again peek into the bytes to find an adapter, so we wrap the cls_stream into a Peekable object.
+        adapter = find_adapter_for_stream(cls_stream)
+        if adapter is None:
+            raise RecordAdapterNotFound("Could not find adapter for file-like object")
+        arg_dict = kwargs.copy()
 
-    mod = importlib.import_module("flow.record.adapter.{}".format(adapter))
+    else:
+        # Either stdout / stdin is given, or a path-like string.
+        url = str(url or "")
+        _, ext = os.path.splitext(url)
 
+        adapter_scheme = ext_to_adapter.get(ext, "stream")
+        if "://" not in url:
+            url = f"{adapter_scheme}://{url}"
+        p = urlparse(url, scheme=adapter_scheme)
+        adapter, _, sub_adapter = p.scheme.partition("+")
+
+        arg_dict = dict(parse_qsl(p.query))
+        arg_dict.update(kwargs)
+
+        cls_url = p.netloc + p.path
+        if sub_adapter:
+            cls_url = sub_adapter + "://" + cls_url
+
+    mod = importlib.import_module("flow.record.adapter.{}".format(adapter))
     clsname = ("{}Writer" if out else "{}Reader").format(adapter.title())
 
     cls = getattr(mod, clsname)
-    arg_dict = dict(parse_qsl(p.query))
-    arg_dict.update(kwargs)
-    cls_url = p.netloc + p.path
-    if sub_adapter:
-        cls_url = sub_adapter + "://" + cls_url
-
     if not out and selector:
         arg_dict["selector"] = selector
 
     if out:
         arg_dict["clobber"] = clobber
 
     log.debug("Creating {!r} for {!r} with args {!r}".format(cls, url, arg_dict))
+
+    if cls_stream:
+        return cls(cls_stream, **arg_dict)
     return cls(cls_url, **arg_dict)
 
 
-def RecordReader(url=None, selector=None, **kwargs):
+def RecordReader(url=None, selector=None, **kwargs) -> AbstractReader:
```

Comment:
Now `RecordReader()` can be called with both a url or a file-like object. If we keep it this way it would be better to also add type hints in the arguments to denote this.

The same applies for `RecordAdapter`. I'm not so sure about the API design, maybe a new wrapper function would be better, eg: `RecordFileReader(fp)` ? @Schamper what do you think?